### ISSUE TITLE
Flakiness in core2.test_pthread_proxying_refcount

### ISSUE THREAD

**Issue by sbc100** [emscripten-core/emscripten#19795]
See https://app.circleci.com/pipelines/github/emscripten-core/emscripten/29174/workflows/e3fe5072-068b-4775-9d62-0a3bbf0c8ad2/jobs/657474

```
AssertionError: JS subprocess failed (/root/emsdk/node/16.20.0_64bit/bin/node --stack-trace-limit=50 --trace-uncaught /tmp/emtest_y_ikdp25/emscripten_test_core2_rv_ej3hp/test_pthread_proxying_refcount.js): 7.  Output:
start
Aborted(Assertion failed: frees_after_cull > frees_before_cull, at: /root/project/test/pthread/test_pthread_proxying_refcount.c,100,main)
/tmp/emtest_y_ikdp25/emscripten_test_core2_rv_ej3hp/test_pthread_proxying_refcount.js:1
var Module=typeof Module!="undefined"?Module:{};var moduleOverrides=Object.assign({},Module);var arguments_=[];var thisProgram="./this.program";var quit_=(status,toThrow)=>{throw toThrow};var ENVIRONMENT_IS_WEB=typeof window=="object";var ENVIRONMENT_IS_WORKER=typeof importScripts=="function";var ENVIRONMENT_IS_NODE=typeof process=="object"&&typeof process.versions=="object"&&typeof process.versions.node=="string";var ENVIRONMENT_IS_PTHREAD=Module["ENVIRONMENT_IS_PTHREAD"]||false;var _scriptDir=[..]
                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                    [..]

RuntimeError: Aborted(Assertion failed: frees_after_cull > frees_before_cull, at: /root/project/test/pthread/test_pthread_proxying_refcount.c,100,main). Build with -sASSERTIONS for more info.
    at abort (/tmp/emtest_y_ikdp25/emscripten_test_core2_rv_ej3hp/test_pthread_proxying_refcount.js:1:7649)
    at ___assert_fail (/tmp/emtest_y_ikdp25/emscripten_test_core2_rv_ej3hp/test_pthread_proxying_refcount.js:1:17949)
    at wasm://wasm/00015f42:wasm-function[23]:0xb6c
    at Module._main (/tmp/emtest_y_ikdp25/emscripten_test_core2_rv_ej3hp/test_pthread_proxying_refcount.js:1:24517)
    at callMain (/tmp/emtest_y_ikdp25/emscripten_test_core2_rv_ej3hp/test_pthread_proxying_refcount.js:1:27699)
    at doRun (/tmp/emtest_y_ikdp25/emscripten_test_core2_rv_ej3hp/test_pthread_proxying_refcount.js:1:28138)
    at run (/tmp/emtest_y_ikdp25/emscripten_test_core2_rv_ej3hp/test_pthread_proxying_refcount.js:1:28306)
    at runCaller (/tmp/emtest_y_ikdp25/emscripten_test_core2_rv_ej3hp/test_pthread_proxying_refcount.js:1:27568)
    at removeRunDependency (/tmp/emtest_y_ikdp25/emscripten_test_core2_rv_ej3hp/test_pthread_proxying_refcount.js:1:7455)
    at /tmp/emtest_y_ikdp25/emscripten_test_core2_rv_ej3hp/test_pthread_proxying_refcount.js:1:13724
```

**Comment by sbc100** [comment]
@tlively

**Comment by sbc100** [comment]
Seems to still be flaky sometimes: https://app.circleci.com/pipelines/github/emscripten-core/emscripten/29266/workflows/c0f92561-28cc-43c9-9fc1-7074df32ede8/jobs/659281

**Comment by tlively** [comment]
I can reproduce this locally. It seems to be the case that Node `Atomics.waitAsync` can take arbitrarily long after `Atomics.notify` to actually resolve its promise. The test assumes that when a `notify` is followed by an atomic write, and that atomic write is observed in the target thread, then the promise will be resolved the next time the target thread returns to its event loop, but that does not seem to be the case.

I don't see any solution to this other than adding a wait back into the test (although we can do it in the `emscripten_async_call` call, which is better than before) or skipping this test on CI. WDYT?

**Comment by tlively** [comment]
No, because the wait is buried in Emscripten library code in `library_pthreads.js`. We wouldn't want to modify it just for this test.

**Comment by RReverser** [comment]
This test has been flakying every PR I made lately at least once, having to do "rerun failed tests" to recover them. Two questions:

1. Can we just disable it until some better resolution is found?
2. @tlively would alternative waiting location on notify state itself (like we discussed recently) possibly fix it? I'd expect so, since in that scenario wait would deterministically return synchronously if notify state has already changed.

**Comment by tlively** [comment]
Disabling the test with a link to this issue sgtm.

Not sure about the second question without paging this all back in...

**Comment by tlively** [comment]
@sbc100, I see this is already skipped if EMTEST_SKIP_FLAKY is true. Should we be setting that flag on CI?

**Comment by sbc100** [comment]
The idea is that we use EMTEST_SKIP_FLAKY on the emscripten-releases CI to keep the builds coming, but that we don't set it on github CI.  At least not yet.    I guess we could but that we might be less motivated to actually fix them?  What do others think?

I guess we could auto-retry them, which is a little better then skipping..

**Comment by RReverser** [comment]
> I guess we could auto-retry them, which is a little better then skipping..

If that's an option, yes please. I get the argument about motivation, but, on the other hand, it's also demotivating to have almost each contribution fail with some unrelated tests. Until there's a proper fix, I'd rather make such contributions easier.

**Comment by sbc100** [comment]
> > I guess we could auto-retry them, which is a little better then skipping..
> 
> If that's an option, yes please. I get the argument about motivation, but, on the other hand, it's also demotivating to have almost each contribution fail with some unrelated tests. Until there's a proper fix, I'd rather make such contributions easier.

I think I agree with this argument... although I've gone back and forth on it over the years.

**Comment by RReverser** [comment]
I mean, the main problem is that the set of people who can fix those tests / underlying functionality has a rather tiny overlap with set of people who receive this error. Which is the main criteria by which it seems to do more harm than good at the moment.

As in, I totally agree it should be fixed over time, I just think it unnecessarily complicates things for external, especially first-time, contributors.

**Comment by sbc100** [comment]
Still seeing this failure.  Just saw it in wasm2js mode.

**Comment by juj** [comment]
This test does repeatedly fail in practically all flavors of the core suite on me, e.g. http://localhost:8010/#/builders/11/builds/310 fails in `core0`, `core2g` and `cores`.

Looking at the test, there is the line

```
  // Cull the zombies! (by forcing a new task queue to be allocated)
  em_proxying_queue* culler = em_proxying_queue_create();
  emscripten_proxy_async(culler, pthread_self(), nop, NULL);

  assert(frees_after_cull > frees_before_cull);
```
which expects that calling `emscripten_proxy_async()` would free the previous proxy queues, and this is not working out.

The test expectation is not right here. Should we delete the test if it is attempting to test something untestable?

What is the test intending to test? I do not find a test intent description anywhere, making me guess what the core focus/purpose of the test is.

What I am confused is that the test filename is `_refcount`, but there are no refcounts anywhere in the test. The only refcounting I find is in `emscripten_thread_mailbox_{un}ref()`.

Is the intent to verify that unrefing a thread mailbox back to zero should close the mailbox? Though then I would have thought the test name to be something like `test_thread_mailbox_refcount`. Though that is probably not it, as that doesn't seem to be what the test is verifying.

Maybe the test purpose was intended be a `test_pthread_proxying_memleak.c` test, to focus that there will be no memory leaks caused by using `emscripten_proxy_async`?

I tried to debug this for a couple of hours, trying a `while(frees_after_cull == frees_before_cull) sched_yield();` type of construct, and also using an `emscripten_set_interval()` to repeatedly check for when the zombies would get culled, but neither of those gives enough room for that to happen. (nothing calls `cull_zombies()` to free anything)

I also tried elevating the `cull_zombies()` loop to outside, and repeatedly calling that in the setInterval() handler, to verify that it does then successfully cull the list, so the memory is reclaimed eventually it looks.

It would be fantastic to resolve this test_pthread_proxying_refcount flakiness, which is one of the most prominent flakes that the suite currently has. (and there are a lot of flakes in the suite, especially around threading)

While debugging this, I was struck by how complex the pthread proxying mechanism has become. Would there be a possibility of removing this concept of zombies altogether to simplify the design?

**Comment by tlively** [comment]
This test checks the task queue lifetime management logic described [here](https://github.com/emscripten-core/emscripten/blob/8e66cb15b6a3ea8b6e2108362d6469e7b96f4eb3/system/lib/pthread/em_task_queue.c#L20-L57). As described there, we must defer freeing task queues to avoid use-after-free bugs when a notification for a task queue arrives after it has been destroyed. We also must allow threads to delegate freeing their task queues to other threads when they shut down to avoid memory leaks when a thread dies before its task queue can be freed. These two requirements require a mechanism like the zombie list, and of course it's important that we have tests for that.

This test is meant to arrange for two task queues to be added to the zombie list and then trigger culling by allocating a third task queue.

To rubber duck it out, let's run through what the test does:

 1. The main thread creates two proxying queues: https://github.com/emscripten-core/emscripten/blob/8e66cb15b6a3ea8b6e2108362d6469e7b96f4eb3/test/pthread/test_pthread_proxying_refcount.c#L70-L73

 2. This allocates two proxying queues, but does not yet allocate any task queue: https://github.com/emscripten-core/emscripten/blob/8e66cb15b6a3ea8b6e2108362d6469e7b96f4eb3/system/lib/pthread/proxying.c#L44-L57

 3. The main thread spawns a worker thread and proxies a task to it on each of the proxying queues: https://github.com/emscripten-core/emscripten/blob/8e66cb15b6a3ea8b6e2108362d6469e7b96f4eb3/test/pthread/test_pthread_proxying_refcount.c#L75-L80
  
 4. `emscripten_proxy_async` [calls](https://github.com/emscripten-core/emscripten/blob/8e66cb15b6a3ea8b6e2108362d6469e7b96f4eb3/system/lib/pthread/proxying.c#L172) `do_proxy`, which [calls](https://github.com/emscripten-core/emscripten/blob/8e66cb15b6a3ea8b6e2108362d6469e7b96f4eb3/system/lib/pthread/proxying.c#L156) `get_or_add_asks_for_thread`. This [creates](https://github.com/emscripten-core/emscripten/blob/8e66cb15b6a3ea8b6e2108362d6469e7b96f4eb3/system/lib/pthread/proxying.c#L103) a task queue, which would cull zombies if there were any at this point (but there are not). Notably this is the only place in which zombies are culled: https://github.com/emscripten-core/emscripten/blob/8e66cb15b6a3ea8b6e2108362d6469e7b96f4eb3/system/lib/pthread/em_task_queue.c#L93-L94

 5. The main thread signals the worker thread to begin executing and waits for the two `executed` flags to be set: https://github.com/emscripten-core/emscripten/blob/8e66cb15b6a3ea8b6e2108362d6469e7b96f4eb3/test/pthread/test_pthread_proxying_refcount.c#L81-L87

 6. The worker thread begins executing: https://github.com/emscripten-core/emscripten/blob/8e66cb15b6a3ea8b6e2108362d6469e7b96f4eb3/test/pthread/test_pthread_proxying_refcount.c#L48-L51

 7. The worker thread executes the proxied tasks on both queues and then destroys the queues: https://github.com/emscripten-core/emscripten/blob/8e66cb15b6a3ea8b6e2108362d6469e7b96f4eb3/test/pthread/test_pthread_proxying_refcount.c#L53-L57

 8. Executing the two proxied tasks sets up two `emscripten_async_call`s to set the two `executed` flags: https://github.com/emscripten-core/emscripten/blob/8e66cb15b6a3ea8b6e2108362d6469e7b96f4eb3/test/pthread/test_pthread_proxying_refcount.c#L43-L45

 9. Destroying the two proxy queues [calls](https://github.com/emscripten-core/emscripten/blob/8e66cb15b6a3ea8b6e2108362d6469e7b96f4eb3/system/lib/pthread/proxying.c#L65) `em_task_queue_destroy` on the two task queues we previously allocated. This is intended to add the two task queues to the zombie list because they have pending notifications: https://github.com/emscripten-core/emscripten/blob/8e66cb15b6a3ea8b6e2108362d6469e7b96f4eb3/system/lib/pthread/em_task_queue.c#L125-L132

 10. The worker thread then calls `emscripten_exit_with_live_runtime` so it can resolve the `emscripten_async_call`s that will set the `executed` flags to signal the main thread to continue executing: https://github.com/emscripten-core/emscripten/blob/8e66cb15b6a3ea8b6e2108362d6469e7b96f4eb3/test/pthread/test_pthread_proxying_refcount.c#L59-L63
 
 11. The main thread allocates a new task queue by proxying to a newly created proxying queue, just as before. This culls the zombies: https://github.com/emscripten-core/emscripten/blob/8e66cb15b6a3ea8b6e2108362d6469e7b96f4eb3/test/pthread/test_pthread_proxying_refcount.c#L94-L96

I suspect the cause of the flake is that sometimes in step 9 the task queues do not have pending notifications, so they do not make it onto the zombie list. We assume there are pending notifications because we assume that the notification telling the worker thread to call into Wasm will arrive before before the notifications about the proxied work. The worker thread does not exit its call into Wasm and return to the event loop to handle the proxied work notifications until after the task queues have been destroyed. However, if it's possible for the proxied work notifications to arrive and be handled before the worker thread calls into Wasm, then there would be no pending notifications. Does that sound possible, @sbc100?

If that is indeed the problem, then it would be solved by waiting on the worker to set a flag saying it has entered Wasm before proxying work to it from the main thread.

**Comment by juj** [comment]
Thanks!
